logger/main.go

```diff
@@ -58,18 +57,31 @@ func main() {
 		setEtcd(client, publishPath+"/drain", drainURI, 0)
 	}
 
-	go client.Watch(publishPath+"/drain", 0, false, drainRespChan, stopChan)
 	go syslogd.Listen(exitChan, cleanupChan, drainChan, fmt.Sprintf("%s:%d", logAddr, logPort))
 	if enablePublish {
 		go publishService(exitChan, client, publishHost, publishPath, strconv.Itoa(logPort), uint64(time.Duration(publishTTL).Seconds()))
 	}
 
+	// HACK (bacongobbler): poll etcd for changes in the log drain value
+	// etcd's .Watch() implementation is broken when you use TTLs
+	//
+	// https://github.com/coreos/etcd/issues/2679
+	go func() {
+		for {
+			resp, err := client.Get(publishPath+"/drain", false, false)
+			if err != nil {
+				log.Printf("warning: could not retrieve drain URI from etcd: %v\n", err)
+				continue
+			}
+			if resp != nil && resp.Node != nil {
+				drainChan <- resp.Node.Value
+			}
+			time.Sleep(time.Duration(publishInterval))
```

Comment:
Apologies for missing to review this. In general, it's better to use a ticker than to sleep. That's because, when exiting, you really should not leave goroutines running. So I would re-write this block as:

```
// HACK (bacongobbler): poll etcd for changes in the log drain value
// etcd's .Watch() implementation is broken when you use TTLs
//
// https://github.com/coreos/etcd/issues/2679
go func() {
  t := time.NewTicker(time.Duration(publishInterval))
  for {
    select {
      case <-exitChan:
        return
      case <-t.C:
        resp, err := client.Get(publishPath+"/drain", false, false)
        if err != nil {
          log.Printf("warning: could not retrieve drain URI from etcd: %v\n", err)
          continue
        }
        if resp != nil && resp.Node != nil {
          drainChan <- resp.Node.Value
        }
    }
  }
}()
```